bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java

```diff
@@ -1040,18 +1043,71 @@ private static boolean checkIsArrayType(BType sourceType, BArrayType targetType,
     }
 
     private static boolean checkIsTupleType(BType sourceType, BTupleType targetType, List<TypePair> unresolvedTypes) {
-        if (sourceType.getTag() != TypeTags.TUPLE_TAG) {
+
+        if (sourceType.getTag() == TypeTags.UNION_TAG) {
+            for (BType memberType : ((BUnionType) sourceType).getMemberTypes()) {
+                if (!checkIsTupleType(memberType, targetType, unresolvedTypes)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        if (sourceType.getTag() != TypeTags.ARRAY_TAG && sourceType.getTag() != TypeTags.TUPLE_TAG) {
+            return false;
+        }
+
+        List<BType> targetTypes = new ArrayList<>(targetType.getTupleTypes());
+        BType targetRestType = targetType.getRestType();
+
+        BTupleType sourceTupleType;
+        if (sourceType.getTag() == TypeTags.TUPLE_TAG) {
+            sourceTupleType = (BTupleType) sourceType;
+        } else {
+            BArrayType sourceArrayType = (BArrayType) sourceType;
+
+            switch (sourceArrayType.getState()) {
+                case UNSEALED:
+                    if (targetRestType == null) {
+                        return false;
+                    } else {
+                        return sourceArrayType.getElementType().getTag() == targetRestType.getTag();
+                    }
+                case CLOSED_SEALED:
+                    if (targetRestType == null && sourceArrayType.getSize() != targetTypes.size()) {
+                        return false;
+                    } else if (targetRestType != null && sourceArrayType.getSize() == targetTypes.size()) {
+                        boolean isSameType = true;
+
+                        for (int i = 0; i < sourceArrayType.getSize(); i++) {
+                            if (sourceArrayType.getElementType().getTag() != targetTypes.get(i).getTag()) {
+                                isSameType = false;
+                            }
+                        }
+                        return isSameType;
+                    }
+                    break;
+            }
+
+            List<BType> fieldTypes = new ArrayList<>();
+            for (int i = 0; i < sourceArrayType.getSize(); i++) {
+                fieldTypes.add(sourceArrayType.getElementType());
+            }
+            sourceTupleType = new BTupleType(fieldTypes);
+        }
+
+        List<BType> sourceTypes = new ArrayList<>(sourceTupleType.getTupleTypes());
+        BType sourceRestType = sourceTupleType.getRestType();
+
+        if ((targetTypes.size() != 0 && (sourceTypes.size() == 0 || sourceTypes.size() != 0))
```

Comment:
Please check the logic here.

⛏️  Code may look simpler if we can use `sourceTypes.isEmpty()` method instead of comparisons like `0` and `!= 0`

♻️ But it looks like after adding rest types to both `sourceTypes` and `targetTypes` and checking the size in (L:1115) make the logic here redundant. 

🔧 Also I think `(sourceTypes.size() == 0 || sourceTypes.size() != 0)` is always true.